### PR TITLE
WIP: Data structure 2.0

### PR DESCRIPTION
**Pull request by xamanu**
This is a new pull request for the overhauled data structure #30 with a lot of wonderful work by @AltNico
It is similar to #89 (closing now), but rebased on #91, which seems to be the next to get in.

### CODE REVIEW

**Comment by xamanu** [comment]
The CI pipeline will fail, because of an error risen by pylint is related to attr.s and can not be fixed: python-attrs/attrs#215

**Comment by xamanu** [comment]
@nlehuby and @prhod, may I request your help on this one? This is a wonderful but deep change and I think it will be difficult to finalize it without your support.

**Status on Accra:**

* It builds (yeah!)
* Accra's unit tests fail because of `File "osm2gtfs/tests/accra_unittests.py", line 151, in test_gtfs_from_cache 'The generated GTFS is different from the expected one')`
* Indeed the GTFS provided for the tests (`accra_tests.zip.ref`) is 1.1M, while the generated one (`accra_tests.zip`) is almost half: 602K

**Status on Fenix:**

* Works until Overpass API complains. No chance to test it further.

**Status on Nicaragua:**

* We use this structure already to generate our GTFSes, but in a different branch without any rebasing. Needs to be further tested.

**Comment by xamanu** [comment]
I changed the cache file names a bit, so this PR can coexists with caching with the old version (before this PR). No need to delete cached files.

**Comment by xamanu** [comment]
In order to make Fenix working properly and to improve the data structure the use of `travel_time` and `duration` was unified and generally improved because all kind of tags of the routes/itineraries are now available in the creators (#76).

File: osm2gtfs/core/routes.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    _itineraries = attr.ib(default=attr.Factory(list))
+
+    def __attrs_post_init__(self):
+        '''
+        Populates the object with information obtained from the tags
+        '''
+        self.name = self.tags['name']
+
+        if "colour" in self.tags:
+            self.route_color = OsmConnector.get_hex_code_for_color(
```
Comment (nlehuby) [inline osm2gtfs/core/routes.py]: you need to import OsmConnector from `osm2gtfs.core.osm_connector`

File: osm2gtfs/creators/accra/trips_creator_accra.py [hunk 2]
```diff
@@ -18,71 +19,83 @@ def add_trips_to_schedule(self, schedule, data):
...
 
             route_index = 0
-            for a_route_ref, a_route in line.routes.iteritems():
-                trip_gtfs = line_gtfs.AddTrip(schedule)
+            itineraries = line.get_itineraries()
+            for a_route in itineraries:
+                trip_gtfs = line_gtfs.AddTrip(feed)
                 trip_gtfs.shape_id = TripsCreator.add_shape(
-                    schedule, a_route_ref, a_route)
+                    feed, a_route.route_id, a_route)
```
Comment (nlehuby) [inline osm2gtfs/creators/accra/trips_creator_accra.py]: you need to replace `a_route.route.id` by `a_route.osm_id` to keep the shape_id identical.

File: osm2gtfs/tests/accra_unittests.py [hunk 3]
```diff
@@ -100,8 +100,8 @@ def test_refresh_routes_cache(self):
         self.assertEqual(len(routes), 277, 'Wrong count of routes in the cache file')
```
Comment (nlehuby) [inline osm2gtfs/tests/accra_unittests.py]: you need to change the name of the cache file here too (`routes-accra` to `accra-routes`)

File: osm2gtfs/tests/accra_unittests.py [hunk 4]
```diff
@@ -117,32 +117,32 @@ def test_refresh_stops_cache(self):
```
Comment (nlehuby) [inline osm2gtfs/tests/accra_unittests.py]: same here

**Comment by xamanu** [comment]
Thanks for the review, @nlehuby. I adjusted the things you mentioned. The whole state is currently still moveing in things from the other branch and make it fit together. Probably you want to wait, until we got it to a stage where we are more confident asking you for help again.

**Comment by xamanu** [comment]
I will close this PR, as it is still not ready, yet, and I don't want to spam you (more) with push notifications. We will re-open, once we are in a shape that we would kindly ask for a review. Thanks.

In case you want to support the ongoing development please have a look on this branch: https://github.com/mapanica/osm2gtfs/tree/data-structure-rebased

**Pull request closed without merging**